xarray/core/rolling.py

```diff
@@ -200,11 +208,61 @@ def _get_keep_attrs(self, keep_attrs):
 
         return keep_attrs
 
+    def _get_rolling_dim_coords(self, all_dims=False) -> Dict[str, Any]:
+        # If any of the dimensions are not padded, the output size can be shorter than the input size
+        # along that dimension, so we also need to shorten the corresponding coordinates.
+
+        # Dimensions which require offsets are those which are not padded, but the logic to determine
+        # the offset is very similar to determining padding sizes.
+        # So, we invert the `pad` flag(s), call `get_pads()`, and work from there.
+
+        coords = self.obj.coords
+        dim = list(self.obj.coords.keys()) if all_dims else self.dim
+        window = self.window
+        center = self.center
+        pad = self.pad
+
+        if pad is False:
+            pad = [True]
+        elif utils.is_list_like(pad):
+            pad = [not p for p in pad]
+
+        offset_pads = utils.get_pads(self.dim, window, center, pad)
+
+        output_coords: Dict[str, Any] = {}
+
+        def get_offsets(d: str):
+            pad_start_offset, pad_end_offset = offset_pads[d]
+
+            start_offset = None if not pad_start_offset else pad_start_offset
+            end_offset = None if not pad_end_offset else -pad_end_offset
+
+            return start_offset, end_offset
+
+        for d in dim:
```

Comment:
Thanks for calling this out, and sorry it isn't clear.

Originally, it was closer to what you suggested here (although I like your version slightly better).

However, we ran into an issue in our own code where we had a second set of coordinates along the dimension without padding--something like
```
In [9]: da
Out[9]: 
<xarray.DataArray (index: 10)>
array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
Coordinates:
  * index    (index) int64 0 1 2 3 4 5 6 7 8 9
    time     (index) int64 1 2 3 4 5 6 7 8 9 10
```

In this case, the second dimension (`time`) is not in `self.dim`, but if `index` needs to be shrunk because of no padding, so does `time`.

So the second part of the loop checks the dimensions of `d` (I'm pretty sure `d` can be multidimensional), to see if any of them are used as rolling dimensions, and reduces `d` along those coordinates as well.

(`test_dataarray.test_rolling_construct()` constructs the `DataArray` above and tests this.)

I think this logic is only needed when returning all coordinates (i.e., when this is being used to construct a new DataArray), and not when the return value is being used by `.sel()`.  Perhaps these should be two separate functions.

Regarding `utils.get_slice_offsets`, I originally wrote that expecting to use it in more than one place, but ended up not needing it (and forgot to remove it).  I believe it's also missing the logic I just described.

Either way, I'll try to clean this up a little to make it clearer, probably with something like you suggest above.

Reply:
Okay, I didn't look closely at your suggestion.  Something close to that should work fine.

Reply:
This is now implemented with something quite similar to your suggested code above (thanks!).